tests/testsuite/build.rs

```diff
@@ -1140,6 +1140,77 @@ fn main(){
         .run();
 }
 
+#[test]
+fn offline_unused_target_dep() {
+    // -Z offline with a target dependency that is not used and not downloaded.
+    Package::new("unused_dep", "1.0.0").publish();
+    Package::new("used_dep", "1.0.0").publish();
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+            [project]
+            name = "foo"
+            version = "0.1.0"
+            [dependencies]
+            used_dep = "1.0"
+            [target.'cfg(unused)'.dependencies]
+            unused_dep = "1.0"
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .build();
+    // Do a build that downloads only what is necessary.
+    p.cargo("build")
+        .with_stderr_contains("[DOWNLOADED] used_dep [..]")
+        .with_stderr_does_not_contain("[DOWNLOADED] unused_dep [..]")
```

Comment:
As a meta note, I personally find "assert does not contain" as a not-very-useful assertion in test suites because there's so many output that don't contain the string. For example if `unused_dep` here had a typo the test would pass but not because we thought it would.

That's where all the original tests from long ago do exhaustive checking of the output which allows for easily detecting bugs/changes. The downside of those tests though is that if you change the output it's really annoying to change the tests, and we have quite a few tests now!

(JFYI)

Reply:
I agree. There is a note in the doc that it is dangerous. Whenever I use it, I try to write the test first so it fails, and then it should only pass when fixed.  However, that doesn't really protect from the future when things change.  I'll leave some notes in #6816.